### DISCUSSION TITLE
Rebase functionality for "Update branch" button

### DISCUSSION THREAD

**Discussion by trivikr** [community/community#3245] [Pull Requests]
**Is your feature request related to a problem? Please describe.**
Right now, GitHub shows "Update branch" button when the base branch is outdated.

![32452897-2117292c-c31b-11e7-8634-6a14c758e591](https://user-images.githubusercontent.com/16024985/106048121-54913f80-6099-11eb-9441-86df5c2b7881.png)

On clicking "Update branch", it creates a merge commit for the PR which merges default branch into PR branch.

![Screen Shot 2021-01-27 at 12 21 57 PM](https://user-images.githubusercontent.com/16024985/106049010-79d27d80-609a-11eb-82f6-756d34ae60aa.png)

It would be nice if this had a "rebase" option, rather than creating a merge commit

**Describe the solution you'd like**
A "rebase" option to update PR branch, which can be configured in `Update branch button` settings which allows choosing between merge and rebase.

This could be similar to existing "Merge button" settings.

![49516468-f3df9100-f899-11e8-8481-c0602ca2ed8c](https://user-images.githubusercontent.com/16024985/106050353-f7e35400-609b-11eb-8aa0-022fea03027b.png)

**Describe alternatives you've considered**
* Stick with the merge commits created by existing "Update branch" button behavior.
* Rebasing from the workspace instead of using "Update branch" button.

**Additional context**
Prior discussion in https://github.com/isaacs/github/issues/1113#issuecomment-403930749

**Comment by evi-liu** [comment]
This space is currently reserved for feedback on Discussions and GitHub for Mobile. This is noted in our [README](https://github.com/github/feedback#github-public-feedback-discussions) and the [pinned discussion](https://github.com/github/feedback/discussions/1).

All other feedback at this time can be shared through our [support channel](https://support.github.com/contact/feedback) or our [community forum](https://github.community/).

**Comment by dimisjim** [comment]
It seems the button was removed altogether today.

Any news regarding that?

**Comment by AndreasPB** [comment]
This is the biggest thing that I miss after moving from GitLab. The `require branches to be up to date` with a merge button is so messy and confusing for collaborators if the project wants to work with a semi-linear strategy.
I'd love to be able to rebase with the button, or just remove it, so people aren't baited into messing up the git history

**Comment by ljharb** [comment]
I am SO EXCITED this landed https://github.blog/changelog/2022-02-03-more-ways-to-keep-your-pull-request-branch-up-to-date/ - but, is there some way i can disable "update branch" while keeping "rebase branch"? Nobody ever wants a merge commit inside a pull request :-)

**Reply by MichaelKetting** [reply]
I'm in the same boat. the next big thing got to be configuration of this feature to only allow rebase-updates. Pretty please!

oh, and please also add this to the Github REST API.

**Reply by mattste** [reply]
Agreed! As much as I try to educate my team, branches still slip through that use merge instead of rebase. It's an easy mistake to make so configuration would be appreciated.

**Reply by masklinn** [reply]
Ditto, there are probably a few cases where merging back into the PR branch is preferable but they ought be few and far between, it should be possible to make "rebase" the default, or even the only update method via the UI. That "merge" is always the default is really not great.

**Reply by BVSokolov** [reply]
Would definitely love to see this change come through as well, i was telling a colleague how I hate that the merge update button is the default

**Reply by jeroenroumen-acc** [reply]
Yes, it's really annoying. Wish there's an option for at least putting rebase as the default.

**Reply by rmacklin** [reply]
I'm not sure if upvoting @ljharb's reply to this post is as likely to get attention on the feature request as upvoting a new/separate post.

There is such a feature request post here: https://github.com/community/community/discussions/12032 so I would encourage folks who want this feature to upvote that as well.

**Reply by rtrad89** [reply]
I would rather want to disable the rebase option and keep the merge commit, because I don't want all signed commits to be unverified.

**Reply by marioslazaro** [reply]
❤️

**Reply by jkester1986** [reply]
would really love for it to be configurable so we can choose which option we want, or at the very least which option to default to. My team uses the rebase strategy, and I'd love to use this feature for branch protection, but I don't want to add it on while the default is only ever merging

**Reply by valentinsa** [reply]
> Estoy MUY EMOCIONADO de que este https://github.blog/changelog/2022-02-03-more-ways-to-keep-your-pull-request-branch-up-to-date/ aterrice, pero, ¿hay alguna manera de deshabilitar "actualizar rama" mientras mantengo "rebase rama"? Nadie quiere una confirmación de fusión dentro de una solicitud de incorporación de cambios :-)

**Comment by kenhuang** [comment]
is there a way to disable `Update with rebase`? as auto rebase via GH would not work with [signed commit](https://docs.github.com/en/authentication/managing-commit-signature-verification).
we would like to only keep `Update with merge commit` option, this will allow squash commit PR to preserve approval status.

🙏

**Reply by ljharb** [reply]
If a squashmerge can keep the signing, why wouldn't a rebase keep it? They're both different SHAs.

**Reply by Thaina** [reply]
> If a squashmerge can keep the signing, why wouldn't a rebase keep it? They're both different SHAs.

Maybe because squashmerge remove all signing to be yours sign only while rebase use other people signing?

**Reply by ManuelRauber** [reply]
> is there a way to disable `Update with rebase`? as auto rebase via GH would not work with [signed commit](https://docs.github.com/en/authentication/managing-commit-signature-verification). we would like to only keep `Update with merge commit` option, this will allow squash commit PR to preserve approval status.
> 
> 🙏

Best would be to have the options selectable by the maintainers, just like we can do it with PR merge strategies

**Reply by rtrad89** [reply]
> > is there a way to disable `Update with rebase`? as auto rebase via GH would not work with [signed commit](https://docs.github.com/en/authentication/managing-commit-signature-verification). we would like to only keep `Update with merge commit` option, this will allow squash commit PR to preserve approval status.
> > 🙏
> 
> Best would be to have the options selectable by the maintainers, just like we can do it with PR merge strategies

I can disable the rebase option when merging a PR, but not when updating a branch. I think it makes sense to disable rebasing everywhere due to the signed commits issue.

**Reply by Pr0methean** [reply]
> Maybe because squashmerge remove all signing to be yours sign only while rebase use other people signing?

Why not sign the merge/rebase commits with GitHub's own key when it's done through the web UI and the sources are all verified?

**Comment by maxwu** [comment]
Cheers for the feature. One question is how to choose email while clicking "update branch"? I have different emails for organizations and personal use.

**Comment by codebydant** [comment]
Hi guys, is there a way to restrict only rebase button? Because if a new contributor does not press the arrow option to select rebase button it will automatically run the "update branch" with the merge functionality.

**Reply by jbnjohnathan** [reply]
Yes I would very much like this option as well

**Reply by matte1** [reply]
Second! This is definitely a pain point for our team!

**Reply by Digirik** [reply]
Wishing for this aswell!

**Reply by mickael-h** [reply]
I see this is still being asked, so I'll join the bandwagon. Please let us restrict to only rebase... Or at least propose rebasing by default...

**Reply by mohamed-haidara-ai71** [reply]
I am joining the bandwagon as well. But since this discussion is answered. Not sure if someone from GH will take a look at this

**Reply by coldfire22x** [reply]
+1

**Reply by erikacarvalho** [reply]
➕1️⃣ this is as necessary as being able to set the PR merge strategy @github

**Reply by zkiezun** [reply]
+1, that would be great to have it!

**Comment by hsuabina** [comment]
I see this conversation started to support an 'Update branch' with an update by rebase option, which is nowadays available. This conversation is considered answered.

I found this other recent conversation (still unanswered) which is specific to allow configuring the possible/default options for the 'Update branch' button, which is what a lot of people in this thread is wishing for.
https://github.com/orgs/community/discussions/57002

**Comment by AndrewEastwood** [comment]
Is there any way to set the default op to be "rebase" rather than just "update" ?

**Comment by fabiendem** [comment]
For those looking for a rebase by default when we hit the Update branch button:

You can go to the Branch protection settings of your repo, create a rule targeting your feature branches (or the ones in which you don't want to see a merge), and enforce the Linear history.

<img width="795" alt="Screenshot 2023-10-20 at 14 11 29" src="https://github.com/community/community/assets/1186282/3f4c80f4-1602-49cc-b382-1373787d0a3d">



You will then get the following:
<img width="333" alt="Screenshot 2023-10-20 at 14 11 05" src="https://github.com/community/community/assets/1186282/24ec18af-b993-4ab8-9b9a-3f0a920748de">

I would bear in mind that your contributors will have to adapt and only use a rebase to update with the main branch locally, which can be a big move for some folks used to the merge.

**Reply by ljharb** [reply]
Unfortunately this doesn't work when you disable rebasemerge and squashmerge (which are not the same as rebase + merge and squash + merge, sadly), so this isn't sufficient.

I should be able to disable the "update with merge commit" button *regardless* of how I've configured the available merge strategies.

**Reply by prenaux** [reply]
That option is broken, if you click "Update branch" (without going through the drop down that shows when you press the arrow) it will create a merge commit even though the "Update with merge commit" is greyed out in the drop down....

**Reply by ryukinix** [reply]
What a mess. I'm very disappointed. GitLab has been doing this well for years.

**Comment by nbolton** [comment]
The [solution](https://github.com/orgs/community/discussions/3245#discussioncomment-7339067) by @fabiendem works, but I'd like to rephrase for the benefit of others, as I missed the point of his comment (that you need a ruleset that specifically tagets the PR branches).

The trick is to have a ruleset for master (include default) and a separate rulset for everything else (all branches excluding default) which forces linear history. That way, the merge button gets disabled and it defaults to rebase.

![image](https://github.com/user-attachments/assets/fe5a2450-f741-4d37-9998-e1f64d69a1a4)

![image](https://github.com/user-attachments/assets/bcae4af7-90cb-4d9f-9a86-f470460274a9)

----

**Edit:** Somehow the following trick of enforcing linear history on non-master/main has stopped working for me, which is frustrating.

![image](https://github.com/user-attachments/assets/1a51da02-a60a-44fe-9122-66c62790a97c)
![image](https://github.com/user-attachments/assets/040ff2ee-e996-47c6-b59e-754126e45fa2)
![image](https://github.com/user-attachments/assets/bacc7da0-fc72-41fd-a93a-d1e2add1236f)

**Reply by jkester1986** [reply]
@nbolton could you clarify exactly what the settings are on both branch rulesets? Is the trick that both rulesets require linear history, and just master needs the "Require branches to be up to date before merging" rule?

**Reply by nbolton** [reply]
I have "Require linear history" set on both rulesets, yes. I also have "Require branches to be up to date before merging" on master.

Here's how I target the Non-master ruleset in case it helps anyone:
![image](https://github.com/user-attachments/assets/e1a26a60-bca5-4cab-addc-7e3037b6cb90)

**Reply by jkester1986** [reply]
thanks!

**Reply by nbolton** [reply]
FWIW, I've noticed that some PRs (maybe older PRs before changing the rulesets) still have the old button behavior. Haven't figured out why yet.

![image](https://github.com/user-attachments/assets/c7227a64-fa8e-418c-ac58-43af144d623d)

Edit: Ah, I think I figured out why. Branch protection rules (e.g. linear history) can only apply to your branches in your repo, i.e. not PRs from external branches. So if you're an open source project accepting PRs from contributor forks, then you'll still see the merge button as you have no control over external repos and their branches.

**Reply by turboBasic** [reply]
so for projects which allow PRs only from own feature branches it should work as expected, right?

**Reply by nbolton** [reply]
> so for projects which allow PRs only from own feature branches it should work as expected, right?

Yeah, if your project does not allow external PRs (i.e. is a private repo) then you should never see a "merge" button; only "rebase".

**Reply by rmacklin** [reply]
Note that there is still this big limitation: https://github.com/orgs/community/discussions/3245#discussioncomment-7343652

> this doesn't work when you disable rebasemerge and squashmerge (which are not the same as rebase + merge and squash + merge, sadly), so this isn't sufficient.
> 
> I should be able to disable the "update with merge commit" button regardless of how I've configured the available merge strategies.

For those who want PRs (which themselves have a linear history) to only be merged via a merge commit (`git merge --no-ff`), they disable the squash and rebase strategies which currently prevents them from checking "require linear history" in any rule. So there's still no way for those repositories to only have the "update with rebase" option on PRs.

**Reply by nbolton** [reply]
Update: This has stopped working for me, frustratingly. I updated my OP to reflect.

**Reply by kragoth** [reply]
> Update: This has stopped working for me, frustratingly. I updated my OP to reflect.

Yep, this has been broken recently. I don't even get the warning message that merge commits are not allowed anymore.... *sigh*

**Reply by sithlord48** [reply]
This just broke on our repo, the update button has returned without any repo setting changes.

**Comment by pdobb-n2** [comment]
I'm seeing this / a similar issue where: Toggling on "Require linear history" for a feature branch directly takes away the option to "Update with rebase" from the "Update branch" drop-down in a PR, leaving only "Update" (which does an "Update with merge commit"). This is backwards since one cannot create Merge commits when "Require linear history" is on.

**Reply by cknd** [reply]
> I'm seeing this / a similar issue where: Toggling on "Require linear history" for a feature branch directly takes away the option to "Update with rebase" from the "Update branch" drop-down in a PR, leaving only "Update" (which does an "Update with merge commit"). This is backwards since one cannot create Merge commits when "Require linear history" is on.

this is still broken, right? a year later

**Comment by nbolton** [comment]
Anyone else always see "Update branch" the first time a PR page loads, then on refresh it says "Rebase branch" instead?

If so, please vote:
- #191285

First load:
<img width="904" height="129" alt="image" src="https://github.com/user-attachments/assets/da21bcb0-9377-4d96-90b0-cd15cd5cdcfb" />

After refresh:
<img width="904" height="129" alt="image" src="https://github.com/user-attachments/assets/75e780a2-c6d9-465b-bc47-fe93243a3311" />
